ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ClasspathComputer.java

```diff
@@ -202,28 +216,32 @@ private static IClasspathAttribute[] getClasspathAttributes(IProject project, IP
 		return attributes;
 	}
 
-	private static void addSourceFolder(IBuildEntry buildEntry, IProject project, HashSet<IPath> paths,
-			ArrayList<IClasspathEntry> result, boolean isTestPlugin) throws CoreException {
+	private static void addSourceFolders(IBuildEntry buildEntry, ClasspathConfiguration context)
+			throws CoreException {
 		String[] folders = buildEntry.getTokens();
+		IProject project = context.javaProject.getProject();
 		for (String folder : folders) {
 			IPath path = project.getFullPath().append(folder);
-			if (paths.add(path)) {
```

Comment:
I assume duplicated tokens are now handled in the very end implicitly since` collectInOriginalOrder()` only uses the first entry with a path?

Reply:
Exactly!